{
  "database_name": "retryable-reads-tests",
  "collection_name": "coll",
  "data": [
    {
      "_id": 1,
      "x": 11
    },
    {
      "_id": 2,
      "x": 22
    },
    {
      "_id": 3,
      "x": 33
    }
  ],
  "minServerVersion": "4.1.6",
  "tests": [
    {
      "description": "db.coll.watch succeeds after InterruptedAtShutdown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 11600
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after InterruptedDueToStepDown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 11602
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after NotMaster",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 10107
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after NotMasterNoSlaveOk",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 13435
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after NotMasterOrSecondary",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 13436
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after PrimarySteppedDown",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 189
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after ShutdownInProgress",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 91
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after HostNotFound",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 7
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after HostUnreachable",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 6
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after NetworkTimeout",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 89
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch succeeds after SocketException",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 9001
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection"
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch fails after two NotMaster errors",
      "clientOptions": {
        "retryReads": true
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 2
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 10107
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection",
        "result": {
          "error": true
        }
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        },
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    },
    {
      "description": "db.coll.watch fails after NotMaster when retryReads is false",
      "clientOptions": {
        "retryReads": false
      },
      "failPoint": {
        "configureFailPoint": "failCommand",
        "mode": {
          "times": 1
        },
        "data": {
          "failCommands": [
            "aggregate"
          ],
          "errorCode": 10107
        }
      },
      "operation": {
        "name": "watch",
        "object": "collection",
        "result": {
          "error": true
        }
      },
      "expectations": [
        {
          "command_started_event": {
            "command": {
              "aggregate": "coll",
              "cursor": {},
              "pipeline": [
                {
                  "$changeStream": {
                    "fullDocument": "default"
                  }
                }
              ]
            },
            "database_name": "retryable-reads-tests"
          }
        }
      ]
    }
  ]
}
